### PR TITLE
[core][Object Store][Bug fix] Fix: ray.put with argument _owner will lost the reference of inner objects

### LINKED ISSUE #30341: [BUG][Object Store] ray.put with argument `_owner` will lost the reference of inner objects
### What happened + What you expected to happen

In function [HandleAssignObjectOwner](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/core_worker.cc#L3355), inner objects are not borrowed by the outer object's owner, will cause fewer references to inner objects than expected.

proposal: borrow inner objects on function [HandleAssignObjectOwner](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/core_worker.cc#L3355)

Do you have any other better solutions? I will fix this bug within a week

### Versions / Dependencies

ray: master, commit: abeb36ecb882919452e7fcc868bfbe876e6319c2

### Reproduction script

``` python
def test_owner_assign_inner_object(ray_start_regular):
    @ray.remote
    class Owner:
        def warmup(self):
            pass

    owner = Owner.remote()
    ray.get(owner.warmup.remote())

    class OutObject:
        def __init__(self, ref):
            self._ref = ref

    inner_ref = ray.put("test")
    out_ref = ray.put(OutObject(inner_ref), _owner=owner)

    del inner_ref
    time.sleep(10)
    # will raise ray.exceptions.GetTimeoutError
    assert ray.get(ray.get(out_ref)._ref, timeout=5) == "test"
```

### Issue Severity

_No response_

### PR DESCRIPTION
**Pull request by Catch-Bull**
## Why are these changes needed?

In function [HandleAssignObjectOwner](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/core_worker.cc#L3355), inner objects are not borrowed by the outer object's owner, will cause fewer references to inner objects than expected.

we need to add nested objects after assigning the owner.

## Related issue number

Closes #30341

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jovany-wang** [comment]
I believe the PR description should be added in details, otherwise the commit message is going to be mess.

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -1055,6 +1055,9 @@ Status CoreWorker::CreateOwnedAndIncrementLocalRef(
                             });
     // Block until the remote call `AssignObjectOwner` returns.
     status = status_promise.get_future().get();
+    // Must call `AddNestedObjectIds` after finished assign owner
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc:1058]: Can you provide some context to this comment and add a link here to the github issue?

Comment (Catch-Bull) [reply src/ray/core_worker/core_worker.cc]: @stephanie-wang sure.

**Comment by Catch-Bull** [comment]
@stephanie-wang Is there anything else that needs to be modified?

**Review by stephanie-wang — APPROVED** [review]

**Comment by stephanie-wang** [comment]
> @stephanie-wang Is there anything else that needs to be modified?

Ah apologies, I thought from the label that you were still working on the PR. Approved! Thanks for fixing this!

**Review by jovany-wang — APPROVED** [review]

**Comment by jovany-wang** [comment]
CI failures are unrelated. They are also failed on master. Merging this...

**Pull request merged**